### PR TITLE
Evidence for contract-key testing

### PR DESCRIPTION
**Pull request by nickchapman-da**
Split contract-key tests from engine tests, and mark for evidence.
- Addresses item 5 of https://github.com/digital-asset/daml/issues/11401

### CODE REVIEW

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ContractKeySpec.scala [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+    new preprocessing.Preprocessor(
+      ConcurrentCompiledPackages(suffixLenientEngine.config.getCompilerConfig)
+    )
+
+  "contract key" should {
+    val now = Time.Timestamp.now()
+    val submissionSeed = crypto.Hash.hashPrivateKey("contract key")
+    val txSeed = crypto.Hash.deriveTransactionSeed(submissionSeed, participant, now)
+
+    // TEST_EVIDENCE: Semantics: contract keys should be evaluated only when executing create
```
Comment (cocreature) [inline daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ContractKeySpec.scala:146]:
```suggestion
    // TEST_EVIDENCE: Semantics: contract keys should be evaluated after ensure clause
```
maybe a bit clearer and matches what this test actually tests afaict.

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ContractKeySpec.scala [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+          seeding = InitialSeeding.TransactionSeed(txSeed),
+        )
+        .consume(_ => None, lookupPackage, lookupKey)
+      result shouldBe a[Right[_, _]]
+    }
+
+    // TEST_EVIDENCE: Semantics: contract keys should be evaluated after ensure clause
+    "be evaluated after ensure clause" in {
+      val templateId =
+        Identifier(basicTestsPkgId, "BasicTests:ComputeContractKeyAfterEnsureClause")
```
Comment (cocreature) [inline daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ContractKeySpec.scala:188]: I’m confused. What is this test testing that the one above doesn’t test?
`ComputeContractKeyWhenExecutingCreate` just calls `ComputeContractKeyAfterEnsureClause`.

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ContractKeySpec.scala [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+        err.message should include(
+          "Update failed due to a contract key with an empty sey of maintainers"
+        )
+      }
+    }
+
+    // Note that we provide no stability for multi key semantics so
+    // these tests serve only as an indication of the current behavior
+    // but can be changed freely.
+    "multi keys" should {
```
Comment (cocreature) [inline daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ContractKeySpec.scala:255]: I think these deserve some markers. Not for the multi-keys behavior (although no harm in already including that) but for the uck mode (we run them in both).

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala [hunk 4]
```diff
@@ -1819,6 +1819,7 @@ class EngineTest
     }
   }
 
+  // TEST_EVIDENCE: Semantics: Contract key behaviour
   "contract key" should {
```
Comment (cocreature) [inline daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala:1823]: did you intend to drop these here? It looks like they are duplicated now.

**Review by cocreature — APPROVED** [review]
Nice, thanks!

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala (hunk 4 above)
Comment (nickchapman-da) [reply daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala]: I totally did mean to drop these! Thanks for catching this. Dropped now.

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ContractKeySpec.scala (hunk 2 above)
Comment (nickchapman-da) [reply daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ContractKeySpec.scala:188]: It looks like the first test shows the behaviour when the `create` is not actually executed:
```
        let _ignore = create ComputeContractKeyAfterEnsureClause with owner
```
I'm not sure how useful it is to test that here, as that's just standard behaviour for a action which is not executed,

**Pull request merged**